pkg/k8s/controllers/reconciler.go

```diff
@@ -0,0 +1,276 @@
+func NewReconciler(cl client.Client, kind, version string, log logrus.FieldLogger, handler Handler) *Reconciler {
+	return &Reconciler{
+		client:  cl,
+		kind:    kind,
+		version: version,
+		log:     log,
+		handler: handler,
+	}
+}
+
+// Kind returns the CRD's kind
+func (r *Reconciler) Kind() string {
+	return r.kind
+}
+
+// Version returns the CRD's version served by this reconciler
+func (r *Reconciler) Version() string {
+	return r.version
+}
+
+// Client returns the client to access k8s API.
+func (r *Reconciler) Client() client.Client {
+	return r.client
+}
+
+// endReconcile is invoked when Reconciler finishes.
+// This method is for logging and metrics and should NOT modify the ReconcileResult.
+func (r *Reconciler) endReconcile(
+	ctx context.Context, //nolint:unparam // Why: ctx might be ignored
+	log logrus.FieldLogger,
+	rr ReconcileResult,
+) {
+	// pass rr by value - EndReconcile should not tamper with the result
+	r.handler.EndReconcile(ctx, log, rr)
+
+	if rr.ReconcileErr != nil {
+		// make sure error messages are never lost
+		log.WithError(rr.ReconcileErr).Error("Reconciler failed to apply the CR")
+		if rr.PropagateErr {
+			log.Error("The error will be propageted to the controller")
+		}
+		return
+	}
+
+	if rr.Skipped {
+		log.Info("Reconciler skipped this event")
+		return
+	}
+
+	log.Info("Reconciler has applied the CR successfully")
+}
+
+// Reconcile is invoked when controller receives resource spec.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.log.WithField("resourceName", req.NamespacedName).WithField("kind", r.Kind())
+	rr := r.doReconcile(ctx, log, req)
+
+	// defer as func to avoid 'capturing' rr - we want to provide latest values to the endReconcile
+	defer func() {
+		r.endReconcile(ctx, log, *rr)
+	}()
+
+	if rr.ReconcileErr != nil {
+		if rr.PropagateErr {
+			// Controlelr infra will retry immediately.
+			return rr.ControllerRes, rr.ReconcileErr
+		}
+
+		if rr.Skipped {
+			// no further action needed. Even if failed - honor Reconcile decision.
+			return rr.ControllerRes, nil
+		}
+
+		// if we do not report error to the controller, we need to requeue for this CR
+		if rr.ControllerRes.Requeue || rr.ControllerRes.RequeueAfter > 0 {
+			// Reconcile alredy asked for requeue, honor it
+			return rr.ControllerRes, nil
+		}
+
+		// Reconcile reported an error and did not ask for a requeue.
+		// This is OK, we will requeue with incrementing intervals.
+		rr.ControllerRes.RequeueAfter = getRequeueDuration(rr.failCount)
+	}
+
+	return rr.ControllerRes, nil
+}
+
+// Reconcile is invoked when resource spec is created or updated.
+func (r *Reconciler) doReconcile(
+	ctx context.Context,
+	log logrus.FieldLogger,
+	req ctrl.Request,
+) *ReconcileResult {
+	rr := ReconcileResult{}
+	in := r.handler.CreateResource()
+	if getErr := r.Client().Get(ctx, req.NamespacedName, in); getErr != nil {
+		log.WithError(getErr).Errorf("unable to get %s CR", r.Kind())
```

Comment:
The error here can also happen if the custom resource is deleted, I believe, but I'd need to test that case to be sure.

Reply:
hmm, then why would controller infra call Reconcile if CR is deleted?

Reply:
It's something to test, I have a vague recollection of this being the case.

> A Kubernetes Controller is a routine running in a Kubernetes cluster that watches for create / update / delete events on Resources

Controllers get `Reconcile` when resources are deleted, so they can cleanup or respond to delete actions.

Reply:
oh, ok - prob needs to handle this particular error somehow by detecting CRD delete case.

Reply:
Let's chat about this offline since I am not sure what is the priority of it. Shall we just ignore the CR for now if deleted since we do not support it anywhere yet anyway?